### PR TITLE
[14.0] [MIG] hr_applicant: Migrated the module in v14.

### PR DESCRIPTION
**Pull request by Chandresh-SerpentCS**
- Module developed by **Pragati Patel**. (Assigned by **Yogesh Mahera**)

### CODE REVIEW

**Comment by Chandresh-SerpentCS** [comment]
@YogeshMahera-SerpentCS Please review the PR. Thanks.

File: hr_applicant/LICENSE/LICENSE [hunk 1]
```diff
@@ -0,0 +1,16 @@
+    OpenERP, Open Source Management Solution
+    Copyright (C) 2011-Today Serpent Consulting Services Pvt. Ltd.
```
Comment (JayVora-SerpentCS) [inline hr_applicant/LICENSE/LICENSE]: 2020-Today

File: hr_applicant/README.rst [hunk 2]
```diff
@@ -0,0 +1,29 @@
+==================
+HR Applicant
+==================
```
Comment (JayVora-SerpentCS) [inline hr_applicant/README.rst]: use = til 't'

File: hr_applicant/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        "views/hr_recruitment_views.xml",
+        "views/hr_recruitment_employee_views.xml",
+        "views/training_views.xml",
+        "views/applicant_profile_report_view.xml",
+    ],
+    "demo": [],
+    "test": [],
+    "installable": True,
+    "auto_install": False,
+    "application": True,
```
Comment (ForamSerpentCS) [inline hr_applicant/__manifest__.py:24]: We may remove "demo", "test" attributes as not any value has been assigned to it. Here, "auto_install" is by default False, so we may remove it.

File: hr_applicant/models/training.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    def onchange_start_date(self):
+        for rec in self:
+            if rec.training_start_date and rec.course_id:
+                end_date = False
+                training_start_date = (
+                    rec.training_start_date
+                    and datetime.datetime.strftime(
+                        rec.training_start_date, DEFAULT_SERVER_DATE_FORMAT
+                    )
+                )
```
Comment (ForamSerpentCS) [inline hr_applicant/models/training.py]: Here, We may optimise this as this version may accept the comparison/assignment directly by datetime object.

File: hr_applicant/models/training.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+    def action_to_be_approve(self):
+        self.write({"state": "to_be_approved"})
+        return True
+
+    def action_approve(self):
+        for rec in self:
+            if not rec.training_attendees:
+                raise ValidationError(
+                    _("Training Attendees should not be Zero!"))
+            rec.write({"state": "approved"})
```
Comment (ForamSerpentCS) [inline hr_applicant/models/training.py]: We can avoid this type of pattern for assignment rather we can use rec.state="approved".

File: hr_applicant/models/training.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+        return True
```
Comment (ForamSerpentCS) [inline hr_applicant/models/training.py]: Here, may no need to return True. Please check for this.

File: hr_applicant/models/training.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+                ]:
+                    raise ValidationError(
+                        _(
+                            "You can not cancel the Training Class if \
+                           all attendees are not in Draft, Awaiting \
+                           Training Start or In complete state!"
+                        )
+                    )
+        self.write({"state": "cancel"})
+        return True
```
Comment (ForamSerpentCS) [inline hr_applicant/models/training.py]: Same as above, we can change the all state assignment and please check for return True if we can remove it.

File: hr_applicant/models/training.py [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+
+    @api.constrains(
+        "class_id", "training_start_date", "training_end_date", "date_of_arrival"
+    )
+    def _check_training_dup(self):
+        # self.ensure_one()
+        for rec in self:
+            if str(rec.training_start_date) < datetime.datetime.now().strftime(
+                DEFAULT_SERVER_DATE_FORMAT
+            ):
```
Comment (ForamSerpentCS) [inline hr_applicant/models/training.py]: Here, we may compare datetime object directly, please check if we can remove string conversion , we may use fields.Datetime.now() rather using python lib datetime.

File: hr_applicant/wizard/select_training.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
+# See LICENSE file for full copyright and licensing details.
+
+from odoo import models, fields, api
+import datetime
```
Comment (ForamSerpentCS) [inline hr_applicant/wizard/select_training.py]: Remove import datetime if not used elsewhere.

File: hr_applicant/wizard/select_training.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
+# See LICENSE file for full copyright and licensing details.
+
+from odoo import models, fields, api
```
Comment (ForamSerpentCS) [inline hr_applicant/wizard/select_training.py]: Import should be in alphabetical order.

**Review by ForamSerpentCS — CHANGES_REQUESTED** [review]

File: hr_applicant/wizard/select_training.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+    training_courses_ids = fields.Many2many(
+        'training.class', string='Training')
+
+    @api.onchange("is_triaing_needed")
+    def onchange_training_courses(self):
+        for rec in self:
+            class_obj = self.env["training.class"]
+            applicant = self.env["hr.applicant"].search(
+                [("id", "=", self._context.get("active_id"))])
```
Comment (ForamSerpentCS) [inline hr_applicant/wizard/select_training.py]: Here, we can define self.env[''] object outside the loop.

**Review by ForamSerpentCS — CHANGES_REQUESTED** [review]

**Comment by pragati-patel98** [comment]
@ForamSerpentCS  I solved the comments, Can you please review it?

**Review by ForamSerpentCS — APPROVED** [review]

**Pull request merged**